storage/store/my_receipts.go

```diff
@@ -0,0 +1,151 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/operation"
+)
+
+// MyExecutionReceipts holds and indexes Execution Receipts.
+// MyExecutionReceipts is implemented as a wrapper around badger.ExecutionReceipts
+// The wrapper adds the ability to "MY execution receipt", from the viewpoint
+// of an individual Execution Node.
+type MyExecutionReceipts struct {
+	genericReceipts     storage.ExecutionReceipts
+	db                  storage.DB
+	cache               *Cache[flow.Identifier, *flow.ExecutionReceipt]
+	indexingOwnReceipts sync.Mutex // lock to ensure only one receipt is stored per block
+}
+
+// NewMyExecutionReceipts creates instance of MyExecutionReceipts which is a wrapper wrapper around badger.ExecutionReceipts
+// It's useful for execution nodes to keep track of produced execution receipts.
+func NewMyExecutionReceipts(collector module.CacheMetrics, db storage.DB, receipts storage.ExecutionReceipts) *MyExecutionReceipts {
+	mr := &MyExecutionReceipts{
+		genericReceipts: receipts,
+		db:              db,
+	}
+
+	store := func(rw storage.ReaderBatchWriter, key flow.Identifier, receipt *flow.ExecutionReceipt) error {
+		mr.indexingOwnReceipts.Lock()
+		rw.AddCallback(func(error) {
+			// not release the lock until the batch is committed.
+			mr.indexingOwnReceipts.Unlock()
+		})
+
+		blockID := receipt.ExecutionResult.BlockID
+		receiptID := receipt.ID()
+		var myOwnReceiptExecutedBefore flow.Identifier
+		err := operation.LookupOwnExecutionReceipt(rw.GlobalReader(), blockID, &myOwnReceiptExecutedBefore)
```

Comment:
This is where we need to pay attention to, the original implementation can check if an own receipt already exist during insert. 
```
			if errors.Is(err, storage.ErrAlreadyExists) {
				var savedReceiptID flow.Identifier
				err := operation.LookupOwnExecutionReceipt(blockID, &savedReceiptID)(tx)
				if err != nil {
					return err
				}
```

The original implementation can do so because it's using badger transaction. However, the new storage abstraction uses badger updates, so we have to check it ourselves. 

The new way it works is to acquire the `indexingOwnReceipts` and ensure no other process is storing own receipts, so that the check for if any own receipts for this block won't be stale.